<!-- Demonstation of obstacle detector -->
<launch>

  <param name="use_sim_time" value="false"/>  
  <arg name="rplidar_flag" default="true" /> <!-- set to true if i want to record-->
  <arg name="hokuyo_flag" default="false" /> <!-- set to true if i want to record-->
  

<node name="rplidarNode1"          pkg="rplidar_ros"  type="rplidarNode" output="screen" if = "$(arg rplidar_flag)">
  <param name="serial_port"         type="string" value="/dev/ttyUSB0"/>
  <param name="serial_baudrate"     type="int"    value="256000"/>
  <param name="frame_id"            type="string" value="laser"/>
  <param name="inverted"            type="bool"   value="false"/>
  <param name="angle_compensate"    type="bool"   value="true"/>
  <remap from="/scan" to="/scan_main"/>
</node>

<!-- 
<node name="scanner_base_to_front_scanner" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 3.141592 -3.141592 base_link laser" if = "$(arg rplidar_flag)"/>
-->
<node name="scanner_base_to_front_scanner" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 3.141592 0 0 base_link laser" if = "$(arg rplidar_flag)"/>

  <node name="scans_merger" pkg="obstacle_detector" type="scans_merger_node">
    <param name="active"            value="true"/>
    <param name="publish_scan"      value="true"/>
    <param name="publish_pcl"       value="true"/>

    <param name="ranges_num"        value="1000"/>

    <param name="min_scanner_range" value="0.30"/>
    <param name="max_scanner_range" value="10.0"/>

    <param name="min_x_range"       value="-10.0"/>
    <param name="max_x_range"       value="10.0"/>
    <param name="min_y_range"       value="-10.0"/>
    <param name="max_y_range"       value="10.0"/>

    <param name="fixed_frame_id"   value="base_link"/>
    <param name="target_frame_id"   value="laser"/>
    <remap from="/front_scan" to="/scan_main"/>
    <remap from="/rear_scan" to="/scan_aux"/>

  </node>


  
  <node name="obstacle_extractor" pkg="obstacle_detector" type="obstacle_extractor_node">
       <param name="active"               value="true"/>
    <param name="use_scan"             value="false"/>
    <param name="use_pcl"              value="true"/>

    <param name="use_split_and_merge"    value="true"/>
    <param name="circles_from_visibles"  value="true"/>
    <param name="discard_converted_segments" value="true"/>
    <param name="transform_coordinates"  value="true"/>

    <param name="min_group_points"     value="3"/>

    <param name="max_group_distance"   value="0.1"/>
    <param name="distance_proportion"  value="0.00628"/>
    <param name="max_split_distance"   value="0.2"/>
    <param name="max_merge_separation" value="0.04"/>
    <param name="max_merge_spread"     value="0.2"/>
    <param name="max_circle_radius"    value="0.6"/>
    <param name="radius_enlargement"   value="0.3"/>

    <param name="frame_id"             value="base_link"/> <!-- map -->
  </node>

  <node name="obstacle_tracker" pkg="obstacle_detector" type="obstacle_tracker_node">
    <param name="active"                  value="true"/>
    <param name="loop_rate"               value="100.0"/>
    <param name="tracking_duration"       value="0.3"/>
    <param name="min_correspondence_cost" value="0.6"/>
    <param name="std_correspondence_dev"  value="0.15"/>
    <param name="process_variance"        value="0.1"/>  
    <param name="process_rate_variance"   value="0.1"/>  
    <param name="measurement_variance"    value="1.0"/>  

    <param name="frame_id"                value="base_link"/> <!-- map -->
    <remap from="tracked_obstacles" to="obstacles"/>
  </node>
  
</launch>
<!-- -->